cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/assembly/KRaftMigrationMockTest.java

```diff
@@ -0,0 +1,434 @@
+                        assertThat(brokerConfig, not(containsString("controller.listener.names")));
+                        assertThat(brokerConfig, not(containsString("controller.quorum.voters")));
+                    }
+                    reconciliation.flag();
+                })));
+    }
+
+    private Kafka kafkaWithStateAndAnno(Kafka current, KafkaMetadataState metadataState, String kraftAnno) {
+        return new KafkaBuilder(current)
+                .editMetadata()
+                    .addToAnnotations(Annotations.ANNO_STRIMZI_IO_KRAFT, kraftAnno)
+                .endMetadata()
+                .withNewStatus()
+                    .withKafkaMetadataState(metadataState.name())
+                .endStatus()
+                .build();
+    }
+
+    static class MockKafkaAssemblyOperator extends KafkaAssemblyOperator {
+
+        public MockKafkaAssemblyOperator(Vertx vertx, PlatformFeaturesAvailability pfa, CertManager certManager, PasswordGenerator passwordGenerator, ResourceOperatorSupplier supplier, ClusterOperatorConfig config) {
+            super(vertx, pfa, certManager, passwordGenerator, supplier, config);
+        }
+
+        @Override
+        ReconciliationState createReconciliationState(Reconciliation reconciliation, Kafka kafkaAssembly) {
+            return new MockReconciliationState(reconciliation, kafkaAssembly);
+        }
+
+        class MockReconciliationState extends ReconciliationState {
+
+            MockReconciliationState(Reconciliation reconciliation, Kafka kafkaAssembly) {
+                super(reconciliation, kafkaAssembly);
+            }
+
+            @Override
+            KafkaReconciler kafkaReconciler(List<KafkaNodePool> nodePools, KafkaCluster kafkaCluster) {
+                return new MockKafkaReconciler(
+                        reconciliation,
+                        kafkaAssembly,
+                        nodePools,
+                        kafkaCluster,
+                        clusterCa,
+                        clientsCa,
+                        config,
+                        supplier,
+                        pfa,
+                        vertx,
+                        kafkaMetadataStateManager
+                );
+            }
+        }
+    }
+
+    static class MockKafkaReconciler extends KafkaReconciler {
+
+        public MockKafkaReconciler(Reconciliation reconciliation, Kafka kafkaCr, List<KafkaNodePool> nodePools, KafkaCluster kafka, ClusterCa clusterCa, ClientsCa clientsCa, ClusterOperatorConfig config, ResourceOperatorSupplier supplier, PlatformFeaturesAvailability pfa, Vertx vertx, KafkaMetadataStateManager kafkaMetadataStateManager) {
+            super(reconciliation, kafkaCr, nodePools, kafka, clusterCa, clientsCa, config, supplier, pfa, vertx, kafkaMetadataStateManager);
+        }
+
+        @Override
+        protected KafkaAgentClient initKafkaAgentClient(Reconciliation reconciliation, Secret clusterCaCertSecret, Secret coKeySecret) {
```

Comment:
Should we use some `AgentClientSupplier` instead of this to make it easier to provide custom implementation?

Reply:
This was exactly on my notes to do. I am happy we are on the same page. I just prefer to not make this change in this PR. It could have more impact and change on other classes, because I would change the ResourceOperatorSupplier. I assume.

Reply:
See other comment.